### PR TITLE
Hrcpp 425/new near cache fo test

### PR DESCRIPTION
**Pull request by rigazilla**
fix for https://issues.jboss.org/browse/HRCPP-425

this is a replacement for
https://github.com/infinispan/cpp-client/pull/323

### CODE REVIEW

**Comment by rigazilla** [comment]
@alanfx The history of #323 was too messy. I've clean up and put everything here.

File: test/NearCacheFailoverTest.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    ~StickyBalancingStrategy() {
+    }
+
+    const transport::InetSocketAddress &getServerByIndex(size_t pos) {
+        const transport::InetSocketAddress &server = servers[pos];
+        return server;
+    }
+private:
```
Comment (alanfx) [inline test/NearCacheFailoverTest.cpp]: Add a blank line here

File: test/NearCacheFailoverTest.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    std::vector<transport::InetSocketAddress> servers;
+    size_t index;
+    const transport::InetSocketAddress &nextServer(
+            const std::set<transport::InetSocketAddress>& failedServers) {
+        if (servers.size() == 1)
+            return servers[0];
+        if (failedServers.size() == 0)
+                {
```
Comment (alanfx) [inline test/NearCacheFailoverTest.cpp]: Just some formatting issues in this file. Sometimes brackets are on the same line, and sometimes not. Plus some spacing issues as well.

File: test/NearCacheFailoverTest.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+        std::string stop_server = python_exec + " " + server_ctl + " stop " + server_pid_filename;
+        std::cout << "Stopping with command: " << stop_server << std::endl;
+        std::system(stop_server.c_str());
+        std::string server_down = python_exec + " " + probe_port + " localhost 11322 60 down";
+        std::cout << "Probing with command: " << server_down << std::endl;
+        std::system(server_down.c_str());
+
+        pre_hits = std::stoi(nearCache.stats()["nearHits"]);
+        // After failover, this get goes remote
```
Comment (alanfx) [inline test/NearCacheFailoverTest.cpp]: Spacing of comments?

**Comment by rigazilla** [comment]
reformatted

File: src/hotrod/impl/NearRemoteCacheImpl.h [hunk 4]
```diff
@@ -99,7 +102,15 @@ class NearRemoteCacheImpl: public RemoteCacheImpl {
             }
             return value;
         }
+        version->version = _nearMap[kbuf].getVersion();
+        return rcb.baseValueUnmarshall(_nearMap[kbuf].getValue());
+    }
+    virtual void stats(std::map<std::string,std::string> &stats) {
```
Comment (alanfx) [inline src/hotrod/impl/NearRemoteCacheImpl.h]: Need a space here after the comma. Also need spaces around the `=` below

File: src/hotrod/impl/operations/AddClientListenerOperation.cpp [hunk 5]
```diff
@@ -42,7 +42,10 @@ void AddClientListenerOperation::releaseTransport(transport::Transport* t) {
...
 }
 
-void AddClientListenerOperation::invalidateTransport(const infinispan::hotrod::transport::InetSocketAddress&, transport::Transport*){
+void AddClientListenerOperation::invalidateTransport(const infinispan::hotrod::transport::InetSocketAddress& addr, transport::Transport* transport){
+    if (transport) {
+        transportFactory->invalidateTransport(addr, transport);
+    }
 }
 
 //
```
Comment (alanfx) [inline src/hotrod/impl/operations/AddClientListenerOperation.cpp]: Can this be removed or do you want to add a comment here?

File: src/hotrod/impl/operations/AddClientListenerOperation.cpp [hunk 6]
```diff
@@ -85,6 +88,7 @@ static void processImmediateEvent(const ClientListener &clientListener, const Co
 
 char AddClientListenerOperation::executeOperation(transport::Transport& transport)
 {
+    this->failed=false;
```
Comment (alanfx) [inline src/hotrod/impl/operations/AddClientListenerOperation.cpp]: Spacing

File: test/NearCacheFailoverTest.cpp [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        return false;
+    }
+    if (rest && rest->compare("value1")) {
+        std::cerr << "Line " << line << " Awaiting value1 got " << (rest ? *rest : "null") << std::endl;
+        return false;
+    }
+    return true;
+}
+
+void releaseRemoteCacheManager(RemoteCacheManager * rcm) {
```
Comment (alanfx) [inline test/NearCacheFailoverTest.cpp]: It seems like you can remove the space after `RemoteCacheManager` here?

File: test/bin/server_ctl.py [hunk 8]
```diff
@@ -37,36 +37,45 @@ def static_java_args(java, jboss_home, config):
     return cmd
 
 def start(args):
-    stop(verbose=False)
     java_exe = args[2]
     ispn_server_home = args[3]
     ispn_server_config = args[4]
+    ispn_server_opts = "-DNOOP";
+    if (len(args)>5) :
```
Comment (alanfx) [inline test/bin/server_ctl.py]: Spacing

File: test/bin/server_ctl.py [hunk 9]
```diff
@@ -37,36 +37,45 @@ def static_java_args(java, jboss_home, config):
...
+        ispn_server_opts = args[5];
+    ispn_server_pid_file = "servers.pkl";
+    if (len(args)>6) :
```
Comment (alanfx) [inline test/bin/server_ctl.py]: Spacing

File: test/bin/server_ctl.py [hunk 10]
```diff
@@ -37,36 +37,45 @@ def static_java_args(java, jboss_home, config):
...
     output.close()
 
     return 0
 
-def stop(verbose=False):
-    if (os.path.exists('servers.pkl')):
-        pkl_file = open('servers.pkl', 'rb')
+def stop(args, verbose=False):
+    ispn_server_pid_file = "servers.pkl";
+    if (len(args)>2) :
```
Comment (alanfx) [inline test/bin/server_ctl.py]: Spacing

**Comment by rigazilla** [comment]
fixed formatting

**Comment by alanfx** [comment]
@rigazilla merging...

**Pull request merged**